IPython/core/async_helpers.py

```diff
@@ -146,28 +146,28 @@ def _should_be_async(cell: str) -> bool:
 
     If it works, assume it should be async. Otherwise Return False.
 
-    Not handled yet: If the block of code has a return statement as  the top
+    Not handled yet: If the block of code has a return statement as the top
     level, it will be seen as async. This is a know limitation.
     """
     if sys.version_info > (3, 8):
         try:
             code = compile(cell, "<>", "exec", flags=getattr(ast,'PyCF_ALLOW_TOP_LEVEL_AWAIT', 0x0))
             return inspect.CO_COROUTINE & code.co_flags == inspect.CO_COROUTINE
-        except SyntaxError:
+        except (SyntaxError, MemoryError):
             return False
     try:
         # we can't limit ourself to ast.parse, as it __accepts__ to parse on
         # 3.7+, but just does not _compile_
         code = compile(cell, "<>", "exec")
-    except SyntaxError:
+    except (SyntaxError, MemoryError):
```

Comment:
I think it better to have a constant variable to hold all these errors, but I don't know where would be good to put it.